conmon-rs/server/src/container_log.rs

```diff
@@ -0,0 +1,84 @@
+use crate::cri_logger::CriLogger;
+use anyhow::{Context, Result};
+use capnp::struct_list::Reader;
+use conmon_common::conmon_capnp::conmon::log_driver::{Owned, Type};
+use std::sync::Arc;
+use tokio::{io::AsyncBufRead, sync::RwLock};
+
+pub type SharedContainerLog = Arc<RwLock<ContainerLog>>;
+
+#[derive(Debug, Default)]
+pub struct ContainerLog {
+    drivers: Vec<LogDriver>,
+}
+
+#[derive(Debug)]
+enum LogDriver {
+    ContainerRuntimeInterface(CriLogger),
+}
+
+#[derive(Clone, Copy, Debug)]
+/// Available pipe types.
+pub enum Pipe {
+    /// Standard output.
+    StdOut,
+
+    /// Standard error.
+    StdErr,
+}
+
+impl ContainerLog {
+    pub fn from(reader: Reader<Owned>) -> Result<SharedContainerLog> {
+        let drivers = reader
+            .iter()
+            .flat_map(|x| -> Result<_> {
+                Ok(match x.get_type()? {
+                    Type::ContainerRuntimeInterface => {
+                        LogDriver::ContainerRuntimeInterface(CriLogger::new(x.get_path()?, None)?)
+                    }
+                })
+            })
+            .collect();
+        Ok(Arc::new(RwLock::new(Self { drivers })))
+    }
+
+    /// Asynchronously initialize all loggers.
+    pub async fn init(&mut self) -> Result<()> {
+        for logger in self.drivers.iter_mut() {
+            match logger {
+                LogDriver::ContainerRuntimeInterface(ref mut cri_logger) => {
+                    cri_logger.init().await.context("init CRI logger")?
```

Comment:
Maybe collect the futures here so we can await them at the same time...

Reply:
might not be worth the effort

Reply:
Hm we would have to spawn a new task for them, right? I mean we could, but maybe we wanna re-consider this when actually having multiple log drivers :grin:

Reply:
let future = cri_logger.init();
futures.push(future);
tokio.join_all(futures).await

Reply:
Something like this... perhaps

Reply:
```rust
diff --git a/conmon-rs/server/src/container_log.rs b/conmon-rs/server/src/container_log.rs
index 82cb83da..335f816f 100644
--- a/conmon-rs/server/src/container_log.rs
+++ b/conmon-rs/server/src/container_log.rs
@@ -56,13 +56,15 @@ impl ContainerLog {
 
     /// Reopen the container logs.
     pub async fn reopen(&mut self) -> Result<()> {
+        let mut futures = vec![];
         for logger in self.drivers.iter_mut() {
             match logger {
                 LogDriver::ContainerRuntimeInterface(ref mut cri_logger) => {
-                    cri_logger.reopen().await.context("reopen CRI logs")?
+                    futures.push(cri_logger.reopen());
                 }
             }
         }
+        tokio::join!(futures::future::join_all(futures));
         Ok(())
     }
 
```

Reply:
@rphillips fine with merging this and following up with that?